admittance_controller/include/admittance_controller/admittance_rule.hpp

```diff
@@ -0,0 +1,309 @@
+// Copyright (c) 2021, PickNik, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+/// \authors: Denis Stogl, Andy Zelenak, Paul Gesel
+
+#ifndef ADMITTANCE_CONTROLLER__ADMITTANCE_RULE_HPP_
+#define ADMITTANCE_CONTROLLER__ADMITTANCE_RULE_HPP_
+
+#include <map>
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+#include "tf2_ros/transform_listener.h"
+#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
+#include <tf2_ros/buffer.h>
+#include "tf2_kdl/tf2_kdl/tf2_kdl.hpp"
+#include "control_msgs/msg/admittance_controller_state.hpp"
+#include "controller_interface/controller_interface.hpp"
+#include "control_toolbox/filters.hpp"
+#include "geometry_msgs/msg/wrench_stamped.hpp"
+#include "config/admittance_controller_parameters.hpp"
+
+// kinematics plugins
+#include "kinematics_interface/kinematics_interface_base.hpp"
+#include "pluginlib/class_loader.hpp"
+
+
+namespace utility {
+// Numerical accuracy checks. Used as deadbands.
+  const double ROT_AXIS_EPSILON = 1e-12;
+}  // utility namespace
+
+namespace admittance_controller {
+
+  struct Transforms {
+    Transforms() {
+      base_control_.setIdentity();
+      base_ft_.setIdentity();
+      base_tip_.setIdentity();
+      world_base_.setIdentity();
+      base_sensor_.setIdentity();
+      base_cog_.setIdentity();
+      base_desired_ft_.setIdentity();
+    }
+
+    Eigen::Matrix<double, 4, 4> base_control_;
+    Eigen::Matrix<double, 4, 4> base_ft_;
+    Eigen::Matrix<double, 4, 4> base_desired_ft_;
+    Eigen::Matrix<double, 4, 4> base_tip_;
+    Eigen::Matrix<double, 4, 4> world_base_;
+    Eigen::Matrix<double, 4, 4> base_sensor_;
+    Eigen::Matrix<double, 4, 4> base_cog_;
+  };
+
+  struct Rotations {
```

Comment:
I don't know what the intention was for having separate `Transforms` and `Rotations` structs. But, obviously a Transform contains a rotation. Maybe you could just delete this `Rotations` struct and use the Transforms everywhere?

Reply:
I somewhat agree. Originally, temporary variables were created for all rotations and transforms used in the calculations. A suggestion was made to store and update them all in one place.

Reply:
I removed the Rotations struct in favor of local variable.